### PR TITLE
Add Database.Servervalue.increment(x)

### PR DESCRIPTION
**Pull request by inlined**
Per offline discussions, it looks like this is pretty simple to add; we just change resolveDeferredValueSnapshot to also accept a cached value.

Added tests for ServeValues in general.

Forgive my lake of familiarity: do we edit packages/database-types directly? While I understand we don't actually want the feature to be exposed directly yet, it seems weird that extending ServerValues in database.ts didn't change database-types/index.ts

### CODE REVIEW

File: packages/database/src/core/util/ServerValues.ts [hunk 1]
```diff
@@ -48,13 +49,62 @@ export const generateWithValues = function(
...
     return value as string | number | boolean;
+  }
+  assert('.sv' in value, 'Unexpected leaf node or priority contents');
+
+  if (typeof value['.sv'] === 'string') {
+    return resolveScalarDeferredValue(value['.sv'], existing, serverValues);
+  } else if (typeof value['.sv'] === 'object') {
+    return resolveComplexDeferredValue(value['.sv'], existing, serverValues);
+  } else {
+    assert(false, 'Unexpected server value: ' + JSON.stringify(value, null, 2));
```
Comment (schmidt-sebastian) [inline packages/database/src/core/util/ServerValues.ts]: Can we just throw an Error here? We don't yet use the asserts from ./util in database.

File: packages/database/test/servervalues.test.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    // server version of the timestamp.
+    expect(values.length).to.equal(2);
+    values.forEach(serverTime => {
+      const delta = Math.abs(serverTime - start);
+      expect(delta).to.be.lessThan(1000);
+    });
+  });
+
+  it('handles increments locally', async () => {
+    // Must go offline because the latest emulator may not support this serer op
```
Comment (schmidt-sebastian) [inline packages/database/test/servervalues.test.ts]: s/sever/server and move comment to line 53

File: packages/database/test/servervalues.test.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        values.push(snap.val() as number);
+      });
+
+      // Because we're offline, we shouldn't await the node operations. This would block the test.
+      node.set(addOne);
+      node.set(5);
+      node.set(addOne);
+
+      node.off('value');
+      expect(values).to.deep.equal([1, 5, 6]);
```
Comment (schmidt-sebastian) [inline packages/database/test/servervalues.test.ts]: Using this test setup it seems relatively straightforward to also test type conversions (both from null and other scalar values). Do you mind adding that?

Comment (inlined) [reply packages/database/test/servervalues.test.ts]: Done

File: packages/database/test/servervalues.test.ts (hunk 2 above)
Comment (inlined) [reply packages/database/test/servervalues.test.ts]: Done

File: packages/database/src/core/util/ServerValues.ts [hunk 4]
```diff
@@ -67,13 +117,18 @@ export const resolveDeferredValue = function(
...
   serverValues: Object
 ): SparseSnapshotTree {
   const resolvedTree = new SparseSnapshotTree();
   tree.forEachTree(new Path(''), function(path, node) {
     resolvedTree.remember(
       path,
-      resolveDeferredValueSnapshot(node, serverValues)
+      resolveDeferredValueSnapshot(
+        node,
+        existing.calcCompleteEventCache(path),
```
[the remark this answers is not in the document]
Comment (inlined) [reply packages/database/src/core/util/ServerValues.ts]: My understanding is that calcCompleteEventCache will return a null node (ChildrenNode.EMPTY_NODE) in this case. I've added a unit (smoke) test with no listeners and an assert against null/undefined.

File: packages/database/src/core/util/ServerValues.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/database/src/core/util/ServerValues.ts]: Oh, sorry, I looked at your PR in an IDE with no diffbase. You can leave it as is.

File: packages/database/src/core/util/ServerValues.ts [hunk 5]
```diff
@@ -67,13 +116,17 @@ export const resolveDeferredValue = function(
  */
 export const resolveDeferredValueTree = function(
   tree: SparseSnapshotTree,
+  syncTree: SyncTree,
   serverValues: Object
 ): SparseSnapshotTree {
   const resolvedTree = new SparseSnapshotTree();
   tree.forEachTree(new Path(''), function(path, node) {
+    const existing = syncTree.calcCompleteEventCache(path);
+    assert(existing != null && typeof existing != 'undefined',
```
Comment (schmidt-sebastian) [inline packages/database/src/core/util/ServerValues.ts]: In Firestore we just do `assert(!!existing)`.

If you want to keep it more explicit, please change to `existing !== null`. `existing != null` already implicitly checks for undefined, but should be rejected by the linter (which we may not have set up correctly).

File: packages/database/src/core/util/ServerValues.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/database/src/core/util/ServerValues.ts]: Do you mind changing the return type of `calcCompleteEventCache()` to ensure that this remains true?

File: packages/database/src/core/util/ServerValues.ts (hunk 1 above)
Comment (inlined) [reply packages/database/src/core/util/ServerValues.ts]: ok

File: packages/database/src/core/util/ServerValues.ts (hunk 5 above)
Comment (inlined) [reply packages/database/src/core/util/ServerValues.ts]: Used !==

**Comment by schmidt-sebastian** [comment]
> Even though you approved the PR is blocked. Do I need to add a second approver?

@Feiyang1 Can you approve the .vscode change?

**Comment by inlined** [comment]
> > Even though you approved the PR is blocked. Do I need to add a second approver?
> 
> @Feiyang1 Can you approve the .vscode change?

Moved to #2365; it would unblock me and seems like a cleaner git history anyway.

File: packages/database/src/core/util/ServerValues.ts [hunk 6]
```diff
@@ -67,13 +116,19 @@ export const resolveDeferredValue = function(
  */
 export const resolveDeferredValueTree = function(
   tree: SparseSnapshotTree,
+  syncTree: SyncTree,
   serverValues: Object
 ): SparseSnapshotTree {
   const resolvedTree = new SparseSnapshotTree();
   tree.forEachTree(new Path(''), function(path, node) {
+    const existing = syncTree.calcCompleteEventCache(path);
```
Comment (schmidt-sebastian) [inline packages/database/src/core/util/ServerValues.ts]: @inlined - Would it be possible to move this call into `resolveComplexDeferredValue`?

Context https://github.com/firebase/firebase-js-sdk/issues/2487
